docs/Using-HashiCorp-Vault.md

```diff
@@ -0,0 +1,84 @@
+
+## About HashiCorp Vault
+
+[HashiCorp Vault](https://www.hashicorp.com/products/vault/) (or just Vault, for short) is just like [HSM](https://en.wikipedia.org/wiki/Hardware_security_module), but in software. It is a convenient mechanism to store secret keys, passwords, or perform cryptographic operations in an isolated secure environment. It provides a convenient API-based interface (RESFtul API) and extremely easy integration with Spring Boot apps via [Spring Cloud Vault](https://cloud.spring.io/spring-cloud-vault). Finally, the enterprise version of Vault supports integration with HSM for even better key protection.
+
+## Installation and Setup
+
+To install Vault, simply follow the [download and installation instruction on HashiCorp website](https://www.vaultproject.io/downloads.html). On Mac, you can install Vault easily by running:
+
+```bash
+$ brew install vault
+```
+
+The easiest way to start Vault for initial testing is by using the development mode with a "zero token", like so:
+
+```bash
+$ vault server --dev --dev-root-token-id="00000000-0000-0000-0000-000000000000"
+```
+
+The Vault starts rather quickly, the last message should be of a format:
+
+```bash
+2019-08-07T20:47:41.280+0200 [INFO]  secrets.kv.kv_3cfd3149: upgrading keys finished
+```
+
+Beware! Never use the "zero token" authentication mentioned above for production environment. Refer to the [Spring Cloud Vault documentation](https://cloud.spring.io/spring-cloud-vault) for more details.
+
+## Adding Database Encryption Key in the Vault
+
+To add a key used to encrypt and decrypt sensitive records in the PowerAuth database, simply call the following command from the terminal:
+
+```bash
+$ vault kv put secret/powerauth-java-server powerauth.server.db.master.encryption.key=[16 bytes encoded in base64, for example 'MTIzNDU2Nzg5MDEyMzQ1Ng==']
+```
+
+Note the key name `secret/powerauth-java-server`. This is the default name of the secure bucket in Vault based on the value of `spring.application.name` property (that is set to `powerauth-java-server` by default). Adjust the name of this key in case you changed the value of `spring.application.name` property. Also, in case you use a custom configuration profile (for example, `testing`), you need to adjust the name by appending the profile name (for example, `secret/powerauth-java-server/testing`).
+
+To check that the value is present in the Vault, you can use:
+
+```bash
+$ vault kv get secret/powerauth-java-server
+``` 
+
+## Configuring PowerAuth Server
+
+In order to make the running PowerAuth Server aware of the running Vault instance and to configure Vault authentication, you need to set the following properties in `bootstrap.properties` (or `bootstrap.yml` file).
+
+```properties
+spring.cloud.vault.enabled=true
+spring.cloud.vault.host=localhost
```

Comment:
Shouldn't we have those also in project resources `bootstrap.properties`?

And from the other side the `spring.cloud.vault.kv.enabled=true` seems to be missing here.

Reply:
Since these are the default Spring properties, we do not need to have them strictly there - they would have default values anyway in the default configuration. Note that there are default values that are good enough (see [Vault source codes](https://github.com/spring-cloud/spring-cloud-vault/blob/master/spring-cloud-vault-config/src/main/java/org/springframework/cloud/vault/config/VaultProperties.java#L43) and that are almost always overridden by production configuration.

Generally, we only specify Spring properties in our `*.properties` files that must have value or have an important configuration meaning for the basic PA stack functionality.

Now, in the `bootstrap.properties` we only specify:

```properties
spring.cloud.vault.enabled=false
spring.cloud.vault.kv.enabled=true
```

The first value is important since this property is by default set to `true`. We do not want to enforce cloud vault to all PA customers (or force them to switch it off).

The second one is important because in case someone enables vault, having this property set to `true` is mandatory. We do not want to suggest disabling it for this reason.